src/combinatorics/additive/salem_spencer.lean

```diff
@@ -0,0 +1,410 @@
+  ... ≤ mul_roth_number s + mul_roth_number t
+      : add_le_add ((hu.mono $ inter_subset_left _ _).le_mul_roth_number $ inter_subset_right _ _)
+          ((hu.mono $ inter_subset_left _ _).le_mul_roth_number $ inter_subset_right _ _)
+
+@[to_additive]
+lemma le_mul_roth_number_product (s : finset α) (t : finset β) :
+  mul_roth_number s * mul_roth_number t ≤ mul_roth_number (s.product t) :=
+begin
+  obtain ⟨u, hus, hucard, hu⟩ := mul_roth_number_spec s,
+  obtain ⟨v, hvt, hvcard, hv⟩ := mul_roth_number_spec t,
+  rw [←hucard, ←hvcard, ←card_product],
+  refine mul_salem_spencer.le_mul_roth_number _ (product_subset_product hus hvt),
+  rw coe_product,
+  exact hu.prod hv,
+end
+
+@[to_additive]
+lemma mul_roth_number_lt_of_forall_not_mul_salem_spencer
+  (h : ∀ t ∈ powerset_len n s, ¬mul_salem_spencer ((t : finset α) : set α)) :
+  mul_roth_number s < n :=
+begin
+  obtain ⟨t, hts, hcard, ht⟩ := mul_roth_number_spec s,
+  rw [←hcard, ←not_le],
+  intro hn,
+  obtain ⟨u, hut, rfl⟩ := exists_smaller_set t n hn,
+  exact h _ (mem_powerset_len.2 ⟨hut.trans hts, rfl⟩) (ht.mono hut),
+end
+
+end monoid
+
+section cancel_comm_monoid
+variables [cancel_comm_monoid α] (s : finset α) (a : α)
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_left :
+  mul_roth_number (s.map $ mul_left_embedding a) = mul_roth_number s :=
+begin
+  refine le_antisymm _ _,
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec (s.map $ mul_left_embedding a),
+    rw subset_map_iff at hus,
+    obtain ⟨u, hus, rfl⟩ := hus,
+    rw coe_map at hu,
+    rw [←hcard, card_map],
+    exact (mul_salem_spencer_mul_left_iff.1 hu).le_mul_roth_number hus },
+  { obtain ⟨u, hus, hcard, hu⟩ := mul_roth_number_spec s,
+    have h : mul_salem_spencer (u.map $ mul_left_embedding a : set α),
+    { rw coe_map,
+      exact hu.mul_left },
+    convert h.le_mul_roth_number (map_subset_map.2 hus),
+    rw [card_map, hcard] }
+end
+
+@[simp, to_additive] lemma mul_roth_number_map_mul_right :
+  mul_roth_number (s.map $ mul_right_embedding a) = mul_roth_number s :=
+begin
+  convert mul_roth_number_map_mul_left s a,
+  ext,
+  exact mul_comm _ _,
+end
+
+end cancel_comm_monoid
+end roth_number
+
+section roth_number_nat
+variables {s : finset ℕ} {k n : ℕ}
+
+/-- The Roth number of a natural `N` is the largest integer `m` for which there is a subset of
+`range N` of size `m` with no arithmetic progression of length 3.
+Trivially, `roth_number N ≤ N`, but Roth's theorem (proved in 1953) shows that
+`roth_number N = o(N)` and the construction by Behrend gives a lower bound of the form
+`N * exp(-C sqrt(log(N))) ≤ roth_number N`.
+A significant refinement of Roth's theorem by Bloom and Sisask announced in 2020 gives
+`roth_number N = O(N / (log N)^(1+c))` for an absolute constant `c`. -/
+def roth_number_nat : ℕ →ₘ ℕ :=
+⟨λ n, add_roth_number (range n), add_roth_number.mono.comp range_mono⟩
+
+lemma roth_number_nat_def (n : ℕ) : roth_number_nat n = add_roth_number (range n) := rfl
```

Comment:
```
@[simp] lemma roth_number_nat_zero :
  roth_number_nat 0 = 0 := rfl
  ```
  would be nice to keep in this file; also this means the proof of `add_roth_number_Ico` can be golfed